### DISCUSSION TITLE
How required are the hooks?

### DISCUSSION THREAD

**Discussion by jez** [arxanas/git-branchless#1286] [Q&A]
The feature of git-branchless I'm most interested in using is `git move` and possibly `git sync`. I'm curious what would happen if I were to delete the `.git/hooks` that `git branchless` sets up. Would these commands still work? Would they work but only in a degraded state?

If it wouldn't work at all, I'd love to be able to read a discussion of what these commands and the hooks themselves do at a high level, so that I can update my own mental model of how `git branchless` works.

Currently I'm assuming that the hooks work in support of features like `git undo` and the ability to track commits not associated with a branch. But if I only plan to use `git move` for the sake of rebasing trees of commits reachable from branches, and `git move` wouldn't be expected to work for this use case without `.git/hooks`, I'd be curious to learn more.

**Comment by arxanas** [comment]
If you haven't already, see [jj](https://github.com/martinvonz/jj) to support similar workflows in a more principled manner.

Besides the `reference-transaction` hook for undo (not critical), git-branchless relies on the `post-rewrite` hook.

- This is invoked by Git for rebases, which is the only way that git-branchless can correlate those rewritten commits.
  - jj handles this by snapshotting branches on each command and inferring the rewrites that happened outside of jj, but this doesn't work if you don't use branches 😭.
- For simplicity of implementation, this is the _only_ way that git-branchless processes rewrites. It actually invokes itself via its own hook.
- In principle, this lets multiple source control tools communicate their rewrites to each other.
  - I spent some time getting jj to invoke the `post-rewrite` hook as well to notify git-branchless, but nothing came of that.

It's possible in principle to have git-branchless notify itself in-memory about rewrites it performed,

- but then you lose the ability to process rewrites from on-disk rebases
  - and besides your own invocations of `git rebase`, git-branchless will initiate its own on-disk rebases, such as if a merge conflict needs to be resolved.
